tensorzero-internal/src/clickhouse/query_builder/mod.rs

```diff
@@ -0,0 +1,1479 @@
+        //  * adds metric columns to the SELECT clause for visibility and debugging
+        let filter_condition_sql = filter_node.to_clickhouse_sql(
+            config,
+            &mut params_map,
+            &mut select_clauses,
+            &mut joins,
+            &mut param_idx_counter,
+        )?;
+        where_clauses.push(filter_condition_sql);
+    }
+
+    let mut sql = format!(
+        r#"
+SELECT
+    {select_clauses}
+FROM
+    {inference_table_name} AS i"#,
+        select_clauses = select_clauses.iter().join(",\n    "),
+        inference_table_name = inference_table_name,
+    );
+
+    if !joins.get_clauses().is_empty() {
+        sql.push_str(&joins.get_clauses().join("\n"));
+    }
+
+    if !where_clauses.is_empty() {
+        sql.push_str("\nWHERE\n    ");
+        sql.push_str(&where_clauses.join(" AND "));
+    }
+    // TODO: add ORDER BY
+
+    if let Some(l) = opts.limit {
+        let limit_param_placeholder = add_parameter(
+            l,
+            ClickhouseType::UInt64,
+            &mut params_map,
+            &mut param_idx_counter,
+        );
+        sql.push_str(&format!("\nLIMIT {limit_param_placeholder}"));
+    }
+    if let Some(o) = opts.offset {
+        let offset_param_placeholder = add_parameter(
+            o,
+            ClickhouseType::UInt64,
+            &mut params_map,
+            &mut param_idx_counter,
+        );
+        sql.push_str(&format!("\nOFFSET {offset_param_placeholder}"));
+    }
+    match opts.format {
+        ClickhouseFormat::JsonEachRow => {
+            sql.push_str("\nFORMAT JSONEachRow");
+        }
+    }
+
+    Ok((sql, params_map))
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum ClickhouseType {
+    String,
+    Float64,
+    Bool,
+    UInt64,
+}
+
+impl Display for ClickhouseType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ClickhouseType::String => write!(f, "String"),
+            ClickhouseType::Float64 => write!(f, "Float64"),
+            ClickhouseType::Bool => write!(f, "Bool"),
+            ClickhouseType::UInt64 => write!(f, "UInt64"),
+        }
+    }
+}
+
+/// Helper to add a parameter and return its SQL placeholder {name:CHType}
+/// The internal_name (e.g. p0, p1) is stored in params_map with its value.
+fn add_parameter<T: ToString>(
+    value: T,
+    ch_type: ClickhouseType,
+    params_map: &mut Vec<QueryParameter>,
+    counter: &mut usize,
+) -> String {
+    let internal_name = format!("p{}", *counter);
+    *counter += 1;
+    params_map.push(QueryParameter {
+        name: internal_name.clone(),
+        value: value.to_string(),
+    });
+    format!("{{{internal_name}:{ch_type}}}")
+}
+
+fn get_select_clauses(
+    function_config: &FunctionConfig,
+    function_name_param_placeholder: &str,
+) -> BTreeSet<String> {
+    let mut select_clauses = BTreeSet::from([
+        format!("{function_name_param_placeholder} as function_name"),
```

Comment:
Can you turn this into a query parameter (just in case we change how `function_name_param_placeholder` is constructed later on)

Reply:
already was!